### PR TITLE
Adds limb sockets, allowing reattachment of robotic limbs and firing of arms

### PR DESCRIPTION
**Pull request by zig-for**
This also lets chainsaw arms and arm blades stay attached after a dismemberment.

### CODE REVIEW

File: code/game/objects/items/weapons/weaponry.dm [hunk 1]
```diff
@@ -1,6 +1,7 @@
 /obj/item/weapon
 
 	var/unique_rename = 0 //allows renaming with a pen
+	var/attach_on_dismember = 0
```
Comment (RemieRichards) [inline code/game/objects/items/weapons/weaponry.dm]: ```FALSE```

File: code/modules/surgery/organs/augments_chest.dm [hunk 2]
```diff
@@ -198,3 +198,77 @@
...
+		return
+
+	var/mob/living/carbon/human/user = caller
+
+	var/turf/T = user.loc
+	var/turf/U = get_step(user, user.dir) // Get the tile infront of the move, based on their direction
+	if(!isturf(U) || !isturf(T))
+		return FALSE
+
+	user.visible_message("<span class='danger'>[user] fires his arm!", "<span class='alertalien'>You fire your arm!.</span>")
```
Comment (RemieRichards) [inline code/modules/surgery/organs/augments_chest.dm]: !.

File: code/game/atoms_movable.dm [hunk 3]
```diff
@@ -126,7 +126,7 @@
 		destination.Entered(src, oldloc)
 		var/area/old_area = get_area(oldloc)
 		var/area/destarea = get_area(destination)
-		if(old_area != destarea)
+		if(destarea && old_area != destarea)
```
Comment (Dorsisdwarf) [inline code/game/atoms_movable.dm]: Is this related to the rest of the PR?

Comment (A-t48) [inline code/game/atoms_movable.dm]: Yes. Sort of? This might only happen on Runtime station, but firing limbs runtimes there, and I figured it must happen for other things.

Comment (A-t48) [inline code/game/atoms_movable.dm]: It looked like get_area could reasonably fail, we have other if checks for it in other places.

File: code/modules/surgery/bodyparts/dismemberment.dm [hunk 4]
```diff
@@ -15,11 +16,53 @@
...
+	if(limb.status == BODYPART_ROBOTIC)
+		stun = 5
+		weaken = 5
+
+	contained_limb = limb
+
+
+/obj/item/projectile/bodypart/on_hit(atom/target, blocked = 0)
+	. = ..()
+	contained_limb.forceMove(src.loc)
```
Comment (RemieRichards) [inline code/modules/surgery/bodyparts/dismemberment.dm]: just ```loc```

**Comment by PKPenguin321** [comment]
So now you can but off a long's armblade and graft it onto somebody else

**Comment by A-t48** [comment]
Just a note: this still needs work for spriting\effects, origin tech, and RD cost. Also need to make it so that if you fire a chainsaw at someone, it does chainsaw attack. :) :)

**Comment by PKPenguin321** [comment]
Lings not longs reeee

**Comment by A-t48** [comment]
@PKPenguin321 only if that ling has a robot arm implant, and the person recieving it has limb socket implants.

**Comment by A-t48** [comment]
I'd be totally fine with dropping the robot arm requirement, too.

**Comment by A-t48** [comment]
Wait, never mind, you are right, with surgury. YES. You can.

**Comment by PKPenguin321** [comment]
You should drop that, yeah. Also it would be cool if an arm that's been dropped with a blade, etc could be used as a weapon

File: code/modules/surgery/bodyparts/dismemberment.dm [hunk 5]
```diff
@@ -83,7 +129,29 @@
...
+
+				// This is kind of horrible and I apologize. 
+				// TODO: allow new_loc on unEquip
+				C.held_items[held_index] = null
+				C.update_inv_hands()
+				if(owner.client)
+					owner.client.screen -= I
+				I.layer = initial(I.layer)
+				I.plane = initial(I.plane)
+				//I.appearance_flags &= ~NO_CLIENT_COLOR
```
Comment (RemieRichards) [inline code/modules/surgery/bodyparts/dismemberment.dm]: don't add commented code, just remove it.

File: code/modules/surgery/bodyparts/dismemberment.dm [hunk 6]
```diff
@@ -83,7 +129,29 @@
...
+
+				did_special_unequip = TRUE
+
+		if(!did_special_unequip)
+			C.unEquip(owner.get_item_for_held_index(held_index), 1)
```
Comment (RemieRichards) [inline code/modules/surgery/bodyparts/dismemberment.dm]: reuse ```I```?

File: code/modules/surgery/organs/augments_chest.dm [hunk 7]
```diff
@@ -198,3 +198,77 @@
...
+	active = FALSE
+
+
+/obj/item/organ/cyberimp/chest/limb_sockets/ui_action_click()
+ 	toggle()
+
+/obj/item/organ/cyberimp/chest/limb_sockets/proc/toggle()
+	var/message
+	if(active)
+		message = "<span class='notice'>You deactivate your arm</span>"
```
Comment (PKPenguin321) [inline code/modules/surgery/organs/augments_chest.dm]: .

File: code/modules/surgery/organs/augments_chest.dm [hunk 8]
```diff
@@ -198,3 +198,77 @@
...
 
+/obj/item/organ/cyberimp/chest/limb_sockets
+	name = "Limb Sockets"
+	desc = "Lets you reattach cyborg limbs and fire your arms!"
+	icon_state = "chest_implant"
+	implant_color = "#AA0000"
+	slot = "limb_sockets"
+	origin_tech = "materials=2;powerstorage=2;biotech=2"
+	actions_types = list(/datum/action/item_action/organ_action/toggle/)
+	var active = FALSE
```
Comment (RemieRichards) [inline code/modules/surgery/organs/augments_chest.dm]: ```var/active = FALSE```, please

**Comment by A-t48** [comment]
@PKPenguin321 for surgury reattachment it doesn't have to be metal, its only if you want quick attach. As for using it as a weapon, should fall out of the work for making fired chainsaw arms do more damage.

**Comment by PKPenguin321** [comment]
This is super cool, I wanted to do something like this for chainsaw arms when dismemberment came out but didn't know how. Big :+1:

File: code/modules/surgery/bodyparts/bodyparts.dm [hunk 9]
```diff
@@ -319,28 +343,22 @@
...
-	var/draw_color
+		var/icon_file = null
 
-	if(skin_tone) //Limb has skin color variable defined, use it
-		draw_color = skintone2hex(skin_tone)
-	if(species_color)
-		draw_color = species_color
-	if(mutation_color)
-		draw_color = mutation_color
+		if(held_index == 2)
```
Comment (RemieRichards) [inline code/modules/surgery/bodyparts/bodyparts.dm]: ```if(held_index % 2 == 0)``` please, for multilimbs.

File: code/modules/surgery/bodyparts/dismemberment.dm [hunk 10]
```diff
@@ -83,7 +129,29 @@
...
+		var/obj/item/I = owner.get_item_for_held_index(held_index)
+		var/did_special_unequip = FALSE
+		if(istype(I, /obj/item/weapon))
+			var/obj/item/weapon/W = I
+			if(W.attach_on_dismember)
+				attached_weapon = W
+				attached_weapon.forceMove(src)
+
+				// This is kind of horrible and I apologize. 
+				// TODO: allow new_loc on unEquip
```
Comment (RemieRichards) [inline code/modules/surgery/bodyparts/dismemberment.dm]: as discussed, *something* should be done about this, but not 100%
Due to issue with items deleting themselves on drop, etc.
you could add a new arg to unEquip, "callDropped = TRUE" and then call unEquip here with that as false, I suppose.

**Comment by jughu** [comment]
Make dismembered chainsaw arms etc not nodrop anymore i mean they are dismembered why is it magically stuck in your hand

**Comment by A-t48** [comment]
@jughu its surgically implanted. You can detach it with a cutting tool.

**Comment by A-t48** [comment]
Had to add an ability for items to override their attack icon, should be useful in the future if you want, say, a tome to have a skull attack icon or something, without having to do special snowflake shit with effects.

**Comment by A-t48** [comment]
Also fixed @PKPenguin321's request - you can now pick up an arm blade arm and attack people with it. There's no inhand sprite though, for any non robot limbs. :(

**Comment by kevinz000** [comment]
code rocket fists :)

**Comment by A-t48** [comment]
@kevinz000 what do you think these are?!

File: code/modules/surgery/bodyparts/bodyparts.dm [hunk 11]
```diff
@@ -59,7 +62,7 @@
 /obj/item/bodypart/attack(mob/living/carbon/C, mob/user)
 	if(ishuman(C))
 		var/mob/living/carbon/human/H = C
-		if(EASYLIMBATTACHMENT in H.dna.species.specflags)
+		if((EASYLIMBATTACHMENT in H.dna.species.specflags) || (user.getorganslot("limb_sockets") != null /* && status == BODYPART_ROBOTIC*/))
```
Comment (RemieRichards) [inline code/modules/surgery/bodyparts/bodyparts.dm]: don't add commented code.

Comment (A-t48) [inline code/modules/surgery/bodyparts/bodyparts.dm]: fix'd

**Comment by A-t48** [comment]
Things left to do -
- [ ] Fix ugly todo code
- [ ] Let attached weapons do damage when thrown
- [ ] Figure out how much damage regular limbs should do
- [ ] Add to R&D

On a separate note, these are really nice if you've been cuffed - it drops the cuffs from your hands, and you can grab your fired limb and run off.

**Comment by PKPenguin321** [comment]
I've always felt that arms and legs should do like at least 9 or 10 brute so you could tear off somebody's arm and beat them to death with it

**Comment by A-t48** [comment]
- [ ] Add new nohands suicide where you fire your head off in your facing direction.

**Comment by optimumtact** [comment]
Is this progressing? Do you mind closing and reopening when you're ready?

**Comment by A-t48** [comment]
It is progressing. I literally committed code for it two days ago. :| I'll probably finish it up tomorrow night, but won't be able to witness test merge because ban.

**Comment by FantasticFwoosh** [comment]
What about that head pop off be a optional deathgasp feature that keeps just your head alive for 60 seconds if you're (head and torso) augged at the neck joint so you have one last chance to knock out your attacker with the force of the collision and a few last noble words.

Wammu style.

**Comment by A-t48** [comment]
That sounds like far too much work for a feature nobody will ever actually use. :P

**Comment by FantasticFwoosh** [comment]
Jokes on you when i self detonate the chemical explosive grenade i have in my detectives fedora with a final act of seppuku flinging my head at you with force decimating the foes that would harm the station.

(ロングキャプテン！) Rongukyaputen!
